text/0000-const-generic-const-fn-bounds.md

```diff
@@ -0,0 +1,684 @@
+changes around the syntax of the language (allowing `const` modifiers in a few places)
+and ensuring that lowering to HIR and MIR keeps track of that.
+The miri engine already fully supports calling methods on generic
+bounds, there's just no way of declaring them. Checking methods for constness is already implemented
+for inherent methods. The implementation will have to extend those checks to also run on methods
+of `impl const` items.
+
+## Precedence
+
+A bound with multiple traits only ever binds the `const` to the next trait, so `const Foo + Bar`
+only means that one has a `const Foo` impl and a regular `Bar` impl. If both bounds are supposed to
+be `const`, one needs to write `const Foo + const Bar`. More complex bounds might need parentheses.
+
+## Implementation instructions
+
+1. Add an `maybe_const` field to the AST's `TraitRef`
+2. Adjust the Parser to support `?const` modifiers before trait bounds
+3. Add an `maybe_const` field to the HIR's `TraitRef`
+4. Adjust lowering to pass through the `maybe_const` field from AST to HIR
+5. Add a a check to `librustc_typeck/check/wfcheck.rs` ensuring that no generic bounds
+    in an `impl const` block have the `maybe_const` flag set
+6. Feature gate instead of ban `Predicate::Trait` other than `Sized` in
+    `librustc_mir/transform/qualify_min_const_fn.rs`
+7. Remove the call in https://github.com/rust-lang/rust/blob/f8caa321c7c7214a6c5415e4b3694e65b4ff73a7/src/librustc_passes/ast_validation.rs#L306
+8. Adjust the reference and the book to reflect these changes.
+
+## Const type theory
+
+This RFC was written after weighing practical issues against each other and finding the sweet spot
+that supports most use cases, is sound and fairly intuitive to use. A different approach from a
+type theoretical perspective started out with a much purer scheme, but, when exposed to the
+constraints required, evolved to essentially the same scheme as this RFC. We thus feel confident
+that this RFC is the minimal viable scheme for having bounds on generic parameters of const
+functions. The discussion and evolution of the type theoretical scheme can be found
+[here](https://github.com/rust-rfcs/const-eval/pull/8#issuecomment-452396020) and is only 12 posts
+and a linked three page document long. It is left as an exercise to the reader to read the
+discussion themselves.
+A summary of the result of the discussion can be found at the bottom of [this blog post](https://varkor.github.io/blog/2019/01/11/const-types-traits-and-implementations-in-Rust.html)
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+* It is not a fully general design that supports every possible use case,
+  but it covers the most common cases. See also the alternatives.
+* It becomes a breaking change to add a new method to a trait, even if that method has a default
+  impl. One needs to provide a `const` default impl to not make the change a breaking change.
+
+# Rationale and alternatives
```

Comment:
This does not mention any alternative to the most hacky part of this proposal, which is `ConstDrop`. I think that warrants a thorough exploration.  You probably did that as part of getting the RFC to this stage, but the arguments for why the most obvious alternative(s) don't work should be collected here.

From my perspective, also in the light of my question about the effect on run-time-callability above, the most immediate choice would have been to make `T: const Drop` have the desired behavior, i.e., make it match any type that can be dropped in const context -- including `u32`.  Basically, `T: Drop` is an entirely useless bound currently; it doesn't let you actually do or assume anything extra and the fact that `u32` does not satisfy it is a surprise.  So why not fix that at least for `T: const Drop`? That *is* a special case, but so is `ConstDrop`, and when comparing two different designs that both need special hacks and inconsistencies (such as `impl ConstDrop` being required for `impl const Drop`) I'd prefer the one that has less moving pieces.
Longer-term, maybe it is also possible to make `T: Drop` have the "sensible" semantics of not posing any restrictions on `T` (so that e.g. `u32` is considered satisfying this bound).

I presume there are reasons why this design was discarded and introducing `ConstDrop` was deemed necessary, but I'd like to see them summarized here.

Reply:
@oli-obk wrote elsewhere

> Having a T: Foo bound on a const fn will allow you to only pass types that implement const Foo if you are within a const context. Drop isn't special here, so you can only call a const fn with a T: Drop bound with a type that has a const Drop impl.

However, given that `Drop` is already special in several ways (you cannot `impl<T: Clone> Drop for Vec<T>`, the `drop` function gets added to any vtable, implementing `Drop` affects move checking for structs, and `Copy` and `Drop` are mutually exclusive), I think it is actually less confusing to make it also special in how `const Drop` relates to `Drop`, as opposed to introducing *another* very magic and special trait, `ConstDrop`.

Basically, `Drop` isn't *really* a trait, and only very superficially behaves like one.  There is value in keeping it as close to the other traits as possible, but that does not seem worth the cost of `ConstDrop`. This is even more so since `T: Drop` is an entirely useless bound -- or is there any case where that is actually useful?

Reply:
Yea, this solution came out of the assumption that `Drop` is just a trait. But you're right, we could just extend `Drop` to cover this case, too. Losening it so `T: Drop` allows `i32`, too is backwards compatible I believe.